### PR TITLE
8293654: Improve SharedRuntime handling of continuation helper out-arguments

### PR DESCRIPTION
**Pull request by shipilev**
(Found this while adapting current mainline to x86_32 port)

After [JDK-8292584](https://bugs.openjdk.org/browse/JDK-8292584), we have `gen_continuation_yield()` that generates compiled entry, and implicitly uses the defaults for other ones (interpreter, exception). We should be more explicit about these, and verify the generators properly initialized all out-parameters.

I think we are only using interpreter/exception entry in `enterContinuation`, but not in `yield`. Notably, `exception_offset` should be `-1` for `nmethod::new_native_nmethod` to treat it as "no exception handlers". 

There a many ways to strengthen this, this PR is the one I like. I can do the symmetric change in aarch64, once we are agree on x86_64 version.

Additional testing:
 - [x] Linux x86_64 fastdebug, `hotspot_loom jdk_loom`
 - [x] Linux x86_64 fastdebug, `tier1`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293654](https://bugs.openjdk.org/browse/JDK-8293654): Improve SharedRuntime handling of continuation helper out-arguments

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [f9101ab7](https://git.openjdk.org/jdk/pull/10241/files/f9101ab7dd0c1674d4ad9f0f02e66fcc3f5acc91)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10241/head:pull/10241` \
`$ git checkout pull/10241`

Update a local copy of the PR: \
`$ git checkout pull/10241` \
`$ git pull https://git.openjdk.org/jdk pull/10241/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10241`

View PR using the GUI difftool: \
`$ git pr show -t 10241`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10241.diff">https://git.openjdk.org/jdk/pull/10241.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp [hunk 1]
```diff
@@ -1602,15 +1600,25 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
...
                              oop_maps,
                              frame_complete,
                              stack_slots,
-                             interpreted_entry_offset,
                              vep_offset);
     } else {
       guarantee(false, "Unknown Continuation native intrinsic");
     }
+
+    if (method->is_continuation_enter_intrinsic()) {
```
Comment (dean-long) [inline src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1611]: I suggest putting these if's inside an #ifdef ASSERT.  Or if you don't mind breaking up the asserts, move them up into the if's above.

**Review by dean-long — APPROVED** [review]

File: src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:1612]: Done the `#ifdef ASSERT` thing in new commit.

**Comment by shipilev** [comment]
As promised, also added the symmetric AArch64 part. `hotspot_loom jdk_loom` passes there, running `tier1` now.

**Comment by shipilev** [comment]
I think this one is pretty simple, but any other Reviews?

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by shipilev** [comment]
Thank you!

/integrate

**Pull request closed without merging**